### ISSUE TITLE
VoiceOver/TalkBack doesn't read the first element when scrolling a PageView

### ISSUE THREAD

**Issue by dannyvalentesonos** [flutter/flutter#53702]
I don't know if this was possibly caused by #44484, but when using 3 fingers to scroll from one PageView's page to another, Voice Over/Talkback doesn't read anything on the page until swiping one finger to the right.
At this point it always reads the second element on the new page, and you have to swipe left to get the first element to read.

What's also weird is that the Accessibility rectangle shows as if the whole PageView container is selected when reading the first element, as opposed to actual element.

The following Flutter code reproduces this problem:

```
PageView(
        children: <Widget>[
          Container(
              color: Colors.blue,
              child: Center(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    children: <Widget>[
                      Text("hello", style: TextStyle(fontSize: 60)),
                      FlatButton(child: Text("open"), onPressed: () => print('pressed'),)
                    ],
                  )
              )
          ),
          Container(
              color: Colors.orange,
              child: Center(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    children: <Widget>[
                      Text("bye", style: TextStyle(fontSize: 60)),
                      FlatButton(child: Text("close"), onPressed: () => print('pressed'),)
                    ],
                  )
              )
          )
        ],
      ),
    )
```

Keep in mind I'm running into this problem in the Add2App scenario on iOS at the moment.

**Comment by dannyvalentesonos** [comment]
@dnfield FYI

**Comment by dannyvalentesonos** [comment]
Another thing I am wondering if it is expected, but Voice Over doesn't let the user know that the view is scrollable either. Are we supposed to add some Semantics node to make that possible?

**Comment by dnfield** [comment]
@darrenaustin has worked on scrolling a11y on iOS recently and may know.  Or @goderbauer

**Comment by dannyvalentesonos** [comment]
Another issue I see also, is that when the entire page view is selected, all text elements in the page get read without any pause.

So if you use this instead:

```
PageView(
        children: <Widget>[
          Container(
              color: Colors.blue,
              child: Center(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    children: <Widget>[
                      Text("hello", style: TextStyle(fontSize: 60)),
                      FlatButton(child: Text("open"), onPressed: () => print('pressed'),)
                    ],
                  )
              )
          ),
          Container(
              color: Colors.orange,
              child: Center(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    children: <Widget>[
                      Text("bye", style: TextStyle(fontSize: 60)),
                      Text("bye", style: TextStyle(fontSize: 60)),
                    ],
                  )
              )
          )
        ],
      ),
    )
```
On the second page, Voice Over will read "Good bye" quickly.

**Comment by dannyvalentesonos** [comment]
Yet another issue I've found, if an Image is present in the tree, Voice over reads the whole page as described above, but says "Image" after reading the text fields first, even if the image is first in the tree.

For example:
```
PageView(
        children: <Widget>[
          Container(
              color: Colors.blue,
              child: Center(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    children: <Widget>[
                      Image.asset("some_image.png"),
                      Text("hello", style: TextStyle(fontSize: 60)),
                      FlatButton(child: Text("open"), onPressed: () => print('pressed'),)
                    ],
                  )
              )
          ),
          Container(
              color: Colors.orange,
              child: Center(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    children: <Widget>[
                      Image.asset("some_image.png"),
                      Text("bye", style: TextStyle(fontSize: 60)),
                      Text("bye", style: TextStyle(fontSize: 60)),
                    ],
                  )
              )
          )
        ],
      ),
    )
```

Having a column with an image, text and text widgets outside of a PageView, Voice Over reads everything as expected.

**Comment by goderbauer** [comment]
> but Voice Over doesn't let the user know that the view is scrollable either

As far as I know this is just how VoiceOver works on iOS. From using VoiceOver on other native apps I have never seen an announcement for this. Usually, the view either just scrolls implicitly when the user moves the a11y focus past what's currently visible or there are other affordances to scroll the view (e.g. a TabBar at the bottom for a PageView, etc).

**Comment by goderbauer** [comment]
There appear to be many different issues discussed in this thread making it really hard to keep track of what the actual issues are. I'll try to respond to some of them:

> What's also weird is that the Accessibility rectangle shows as if the whole PageView container is selected when reading the first element, as opposed to actual element.

This one can be fixed by wrapping the Container in a `Semantics(explicitChildNodes: true, child: Container(...))`. Or by wrapping the Text widget in `Semantics(container: true, child: Text(...))`. 

The `Text` widget tries to add its semantics information to the furthest possible ancestor, which in this case is covering the entire screen. With the settings above, the semantics information will be contributed more locoally.

**Comment by goderbauer** [comment]
> At this point it always reads the second element on the new page, and you have to swipe left to get the first element to read.

On iOS if a previously focused node is removed, iOS more-or-less randomly picks a new node to focus (The new node is usually close to the location of the old node). If you want to always focus the first node, you can pretend that the new page represents a new route by wrapping you Container in `Semantics(scopesRoute: true, child: Container(...)`. A new route will always have its focus placed on the first child on screen (and a chime sound plays to tell users that they are on a new route). This appears to be what Apple is doing on the Home screen of iOS as well (which is similar to a PageView).

**Comment by goderbauer** [comment]
> Another issue I see also, is that when the entire page view is selected, all text elements in the page get read without any pause.

This is also resolved by wrapping the Container in a Semantics(explicitChildNodes: true, child: Container(...)). Or by wrapping the Text widgets individually in Semantics(container: true, child: Text(...)).

Text widgets always try to merge their content with their siblings into one. If you don't want that, use the code above.

**Comment by dannyvalentesonos** [comment]
> There appear to be many different issues discussed in this thread making it really hard to keep track of what the actual issues are. I'll try to respond to some of them:
> 
> > What's also weird is that the Accessibility rectangle shows as if the whole PageView container is selected when reading the first element, as opposed to actual element.
> 
> This one can be fixed by wrapping the Container in a `Semantics(explicitChildNodes: true, child: Container(...))`. Or by wrapping the Text widget in `Semantics(container: true, child: Text(...))`.
> 
> The `Text` widget tries to add its semantics information to the furthest possible ancestor, which in this case is covering the entire screen. With the settings above, the semantics information will be contributed more locoally.

Thanks @goderbauer We did figure that out last night and tried that, though it has some other weird effects, which we're working through. But I'm unclear why the Image gets read out of order (from one of my comments above).

**Comment by dannyvalentesonos** [comment]
> > At this point it always reads the second element on the new page, and you have to swipe left to get the first element to read.
> 
> On iOS if a previously focused node is removed, iOS more-or-less randomly picks a new node to focus (The new node is usually close to the location of the old node). If you want to always focus the first node, you can pretend that the new page represents a new route by wrapping you Container in `Semantics(scopesRoute: true, child: Container(...)`. A new route will always have its focus placed on the first child on screen (and a chime sound plays to tell users that they are on a new route). This appears to be what Apple is doing on the Home screen of iOS as well (which is similar to a PageView).

Thanks @goderbauer . I've tried that, but it doesn't work on iOS.
I've logged #53755 for this.

**Comment by goderbauer** [comment]
> But I'm unclear why the Image gets read out of order (from one of my comments above).

iOS always reads out the fact that something is an image as the last information. In your example, the Text widgets and the Image widgets are merged into one Node. Because of iOS's read-out order, the labels from the text widgets are read first, followed by the image flag. If you use `Semantics(explicitChildNodes: true, ..)` around the container, the image will get its own semantics node and you will not have that problem anymore. You should also give the Image a semantics label or wrap it in ExcludeSemantics if you don't want it to be part of the tree.

**Comment by goderbauer** [comment]
Can we close this issue? It seems like we have https://github.com/flutter/flutter/issues/53755 on file for your remaining issues. And the other things don't appear actionable from Flutter's perspective. You'll just have to use `Semantics(explicitChildNodes: true, ..)` to get the a11y tree into the right structure.

**Comment by dannyvalentesonos** [comment]
> Can we close this issue? It seems like we have #53755 on file for your remaining issues. And the other things don't appear actionable from Flutter's perspective. You'll just have to use `Semantics(explicitChildNodes: true, ..)` to get the a11y tree into the right structure.

Yes I am ok with that. I appreciate all the responses as it helps immensely. Thanks!

**Comment by dannyvalentesonos** [comment]
> > Another issue I see also, is that when the entire page view is selected, all text elements in the page get read without any pause.
> 
> This is also resolved by wrapping the Container in a `Semantics(explicitChildNodes: true, child: Container(...))`. Or by wrapping the Text widgets individually in `Semantics(container: true, child: Text(...))`.
> 
> Text widgets always try to merge their content with their siblings into one. If you don't want that, use the code above.

@goderbauer This solution works, but causes a lot of crashing. I logged #53827
